### PR TITLE
Create release_notes and package_manifest automatically

### PR DESCRIPTION
**Pull request by justincely**
addresses #13 

Hi-jacking "python setup.py install" command to run release_notes.py and package_manifest.py which will create release_notes.rst and package_manifest.rst.  This change will allow RTD to install the package, create the files, and have docs that update each time the astroconda package is updated.

Note that release_notes.py uses the github API to pull releases from spacetelescope, and un-authorized github queries are limited to 60 calls/hour.  Any more than this will generate an error.  As long as we don't try to build once a minute this should be fine.

### CODE REVIEW

**Comment by pllim** [comment]
For Sphinx to auto-generate doc, hacking `setup.py` itself is a bit of overkill IMHO. I [suggested something](https://github.com/astroconda/astroconda/issues/13#issuecomment-229527699), but if that is not feasible, at most hacking `conf.py` should be enough?

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import os
+from setuptools import setup, find_packages
+from setuptools.command.install import install
+
+class MyInstall(install):
+    def run(self):
+        install.run(self)
+        os.system('python source/release_notes.py')
```
Comment (pllim) [inline setup.py]: Does this affect any future plan for Windows support?

Comment (jhunkeler) [inline setup.py]: I'm think `release_notes.py`'s dependencies are cross-platform, but I've never tried running it under Windows so I could be lying.

Comment (justincely) [inline setup.py]: I added in some things to make paths platform-independent, but haven't tested any more of it.  Will definitely need to do if I get the base idea working

**Comment by justincely** [comment]
@pllim honestly, I wanted to try it out and see how it worked.  The automodule idea could definitely work, but creating dummy modules to create the docs actually seemed more confusing to me than having setup.py generate them.  Same argument with conf.py.  I've only seen conf.py set configurations for RTD, not actively create documentation.  

Regardless, the bigger issue is the github API only allows 60 calls/hour from an anonymous login.  So regardless of where the creation sits, i need to reduce the API calls in the doc generating scripts so that it can actually complete, which i'm working on doing.  After that, I'm happy to re-asses where the actual doc-creation stuff goes.

**Comment by pllim** [comment]
If you ever want to find out how to use `conf.py` to generate stuff, there is an example in `stginga`.

**Comment by justincely** [comment]
thanks, i'll look into that.  one problem with my current solution is that it can fail silently in travis, so i may have to go with one of your suggestions

**Comment by pllim** [comment]
Let me know if you have questions. @eteq takes the credit for the auto notebook doc generation for `stginga`. Hope one of the ideas works for you. Good luck!

**Comment by pllim** [comment]
Idea: If somehow `astroconda-contrib` can be a submodule of this, then maybe you can avoid github access. All you need is submodule update and then grab version info from individual package `meta.yml`.

**Comment by justincely** [comment]
@pllim the version numbers aren't the issue - those are pulled straight from the metapackages on ssb.stsci.edu.  Rather, @jhunkeler set it up to scrape all the release notes of all spacetelescope repos and collate them into a release_notes.rst page.  This requires pinging the github API more than the allowed 60 times/hour under an unauthorized connection.

We tried creating a read-only token and commiting it to the repo so any calls could have restricted authorization, but github immediately noticed and revoked the token, so that's out too.  

I think this may just be a bust. I'm gonna ping RTD though and ask - they clearly already have a token because we've granted access for the webhook, if it's available at all in the container maybe that can be a solution.  Otherwise my next best bet would be to create a webhook into astroconda-contrib, such that everytime a release is made a listener pulls down astroconda, re-creates the docs, and pushes back up.  Kinda excessive, and not high priority, but that's all i can think of right now.

**Comment by pllim** [comment]
Seems ineffective to pull out release note like that. Release note, once released, does not change anymore. It seems more effective for AstroConda to store them when a release is made somewhere static (e.g., can be a text file of some sort) and then just pull from that static file when you want a release note. To kick things off, this static file can be manually created, and then automatically updated from then on. Does this make sense? Just an idea.

**Comment by jhunkeler** [comment]
@pllim If I'm reading this right, you're saying we should generate the release notes and host them remotely, and have RTD pull them at build-time?

**Comment by pllim** [comment]
Well, my idea (might or might be feasible to you):

1. Someone submit a pull request to `astroconda-contrib` with a new release.
2. After it is merged, AstroConda performs some build magic (I am really oblivious to this part). During this magical time, AstroConda is aware of what had a new release, hence it grabs that release note and stores it somewhere (remote or local is up to you); let's call this "somewhere" *Narnia*.
3. `astroconda` RTD then only has to grab notes from *Narnia* (one connection) rather than from many GitHub repos separately. How does it know when to grab? That depends on where *Narnia* is -- if it is local to `astroconda`, then it kicks off when needed via webhook; if it is remote, then a nightly cron job should do it.

**Comment by pllim** [comment]
@jhunkeler , do you really need the release notes text? It would be much easier to just add links that point to URLs pointing to the notes.

**Comment by justincely** [comment]
@pllim Yeah, i think think this will have to be a manual solution for now.  Either putting links, or just remembering to quick run the script with each update. 

sorry for the false hope @jhunkeler, I tried to automate it and failed!

**Comment by jhunkeler** [comment]
@pllim @justincely  Including release notes in the documentation was a requirement passed down to me from above. Instead of aggregating the release notes text with the API, we could simply point people to the github "Release" page for each repo.

Or do you think a list of URLs too ugly?

**Comment by pllim** [comment]
Personally as a user, I prefer links. I don't need to read release notes that I don't care about. If I care enough about a package, I'll just click the link.

**Comment by justincely** [comment]
@jhunkeler I think i agree with @pllim.  It's actually easier to parse just a couple links to the release notes of the package the user cares about than scrolling through the text.  From the requirements that were handed down to you - would this be considered OK?  Or was the desire to have them all in one place?

**Comment by justincely** [comment]
It only takes a single API hit to auto-pull all the repos - so i think I could auto-generate the page of links with anonymous authentication

**Comment by justincely** [comment]
Ok, i think this is ready to go.  @jhunkeler, i removed the pandoc install and usage, since it seemed to be creating MD and then instantly converting to RST.  So now it just creates RST from the start.  If this is a problem for anything else, let me know, but this should work.

May need a couple iterations on RTD as well - since it doesn't test pull requests.

**Comment by pllim** [comment]
Do you have a locally generated HTML? If so, it would be nice to see what the end product look like first before merging. Also, to keep `master`'s history clean, I would recommend a "squash and merge".

**Comment by justincely** [comment]
@pllim, yeah i do.  Is there a good way to share that through github - or just stick it on central store?

**Comment by pllim** [comment]
Central store should be sufficient. Or if you know how, you can even have our branch site host it.

**Comment by justincely** [comment]
@pllim @jhunkeler, I've squashed the commits down into just one.  

You can see a local copy of the docs at /user/ely/temp/astroconda/.  The pages that have changed, and will be auto-generated, are package_manifest.html and release_notes.html.

**Comment by pllim** [comment]
@justincely , thanks. I notice some weird stuff in the doc:

* Release notes include links to packages not in AstroConda yet (e.g., `synphot_refactor`) and also those that are removed (e.g., `astrolib.coords`, `pyfits`). The notes should only contain packages that are actually distributed. (Some links don't really have actual release notes because the developer didn't write any with the release but that is not our problem.)
* In the package listing, is it normal to list the same package over and over again but including older versions? E.g.

```
osx-64 packages

* acstools-2.0.0
* acstools-2.0.2
* acstools-2.0.4
* acstools-2.0.5
```

Also, I think it is cleaner to separate sub-sections in "Packages" into sub-pages. With the current layout, it is hard to see all the available builds and metapackages.

**Comment by justincely** [comment]
@pllim good point on the packages not part of astroconda.  I've added a bit to not include release notes from packages not distributed in astroconda.  @jhunkeler, note that this does change from previous behavior - can i get your review of this as well?

For the multiple package listings - this is OK.  Each metapackage manifest will list only 1 version, but each platform manifest (linux-64, osx-64) will list multiple since we distribute multiple versions as installable.  On your broader point of re-organizing the package manifest pages - i think it's a good point, but it requires a little bit more thought as to how we want the final version to look.  There's multiple versions of each package (diff delivery, diff numpy, diff python), and just sticking each into a new subpage may not be enough to clarify it sufficiently.  I'd like to move this cleanup to a new issue, as this one was more focused on just making the current stuff more automatic -- unless you have objections?

**Comment by justincely** [comment]
and you can see a new local build of the pages, with the release notes cleanup, at the same location as before

**Comment by pllim** [comment]
I am okay with re-org as new issue. Good work with the automation! @jhunkeler should have the final review and merge.

**Comment by pllim** [comment]
One more thing, you should add the *generated* files to `.gitignore` so they don't get accidentally added back to version tracking.

**Comment by justincely** [comment]
right thanks, done.   -  @jhunkeler ready for review and merge!

**Comment by jhunkeler** [comment]
@justincely 
Great work! Thanks... This is definitely a major improvement over what we had before.

Do you think we need to add a little blurb to the top of the page such as, "some packages may not contain release notes"? I feel like its self-explanatory if no release data shows up after clicking on a link, but it may not be apparent to some.

**Comment by pllim** [comment]
@jhunkeler , :+1: on disclaimer.

**Comment by justincely** [comment]
i just put in a little blurb to that effect

**Comment by jhunkeler** [comment]
🎅

**Pull request merged**